libfeed/feed-knowledge-app-card-store.c

```diff
@@ -0,0 +1,297 @@
+      break;
+    case PROP_THUMBNAIL_SIZE:
+      g_value_set_uint (value, priv->thumbnail_size);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eos_discovery_feed_knowledge_app_card_store_finalize (GObject *object)
+{
+  EosDiscoveryFeedKnowledgeAppCardStore *store = EOS_DISCOVERY_FEED_KNOWLEDGE_APP_CARD_STORE (object);
+  EosDiscoveryFeedKnowledgeAppCardStorePrivate *priv = eos_discovery_feed_knowledge_app_card_store_get_instance_private (store);
+
+  g_clear_pointer (&priv->title, g_free);
+  g_clear_pointer (&priv->uri, g_free);
+  g_clear_pointer (&priv->synopsis, g_free);
+  g_clear_pointer (&priv->bus_name, g_free);
+  g_clear_pointer (&priv->knowledge_search_object_path, g_free);
+  g_clear_pointer (&priv->knowledge_app_id, g_free);
+}
+
+static void
+eos_discovery_feed_knowledge_app_card_store_dispose (GObject *object)
+{
+  EosDiscoveryFeedKnowledgeAppCardStore *store = EOS_DISCOVERY_FEED_KNOWLEDGE_APP_CARD_STORE (object);
+  EosDiscoveryFeedKnowledgeAppCardStorePrivate *priv = eos_discovery_feed_knowledge_app_card_store_get_instance_private (store);
+
+  g_clear_object (&priv->thumbnail);
+}
+
+static void
+eos_discovery_feed_knowledge_app_card_store_init (EosDiscoveryFeedKnowledgeAppCardStore *store)
+{
+}
+
+static void
+eos_discovery_feed_knowledge_app_card_store_class_init (EosDiscoveryFeedKnowledgeAppCardStoreClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->constructor = eos_discovery_feed_knowledge_app_card_store_constructor;
+  object_class->get_property = eos_discovery_feed_knowledge_app_card_store_get_property;
+  object_class->set_property = eos_discovery_feed_knowledge_app_card_store_set_property;
+  object_class->dispose = eos_discovery_feed_knowledge_app_card_store_dispose;
+  object_class->finalize = eos_discovery_feed_knowledge_app_card_store_finalize;
+
+  eos_discovery_feed_knowledge_app_card_store_props[PROP_TITLE] =
+    g_param_spec_string ("title",
+                         "Title of Card",
+                         "The title of the card",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_card_store_props[PROP_URI] =
+    g_param_spec_string ("uri",
+                         "URI of Content",
+                         "The unique URI for the content in the app",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_card_store_props[PROP_SYNOPSIS] =
+    g_param_spec_string ("synopsis",
+                         "Synopsis of Content",
+                         "Brief description of content",
+                         "",
+                         G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY);
+
+  eos_discovery_feed_knowledge_app_card_store_props[PROP_THUMBNAIL] =
+    g_param_spec_object ("thumbnail",
```

Comment:
I'm not sure why this needs to be an input stream. Could it instead either be a path/URI, and this class load the stream, or an already-loaded icon?

Reply:
Ah, looks like my comment here didn't get saved.

It needs to be an input stream since that's what we get from `EosShardBlob`. Since the image can come from either a path or a shard blob, this is the common abstraction for both of them. Also, it is what `widgetSurfaceCache.js` uses.